treewide: migrate buildDotnetModule update scripts to use fetch-deps

IvarWithoutBones opened the pull request:
###### Description of changes
This migrates most of the update scripts for `buildDotnetModule` derivations to use `passthru.fetch-deps` to regenerate their nuget lockfiles, instead of doing it manually. This should be a lot more reliable as the version of dotnet, flags passed to `dotnet restore`, and project file locations are inherited from the parent derivation.

~~The following update scripts are still running `nuget-to-nix` manually (found with `rg "nuget-to-nix" -g "*.sh" ./pkgs`):~~
~~* `./pkgs/applications/misc/ArchiSteamFarm/update.sh`~~
~~This fetches dependencies in a cross compiling environment, needs testing~~
~~* `./pkgs/servers/jellyfin/update.sh`~~
    ~~Runs `dotnet restore` for a multitude of RIDs, needs testing for all targets.~~

~~Note that I do plan to migrate those, but `vid-stab` is currently broken which causes the aarch64 cross env to fail. Because of that I cannot test these right now.~~


These are all migrated now.

Related PR: https://github.com/NixOS/nixpkgs/pull/187150. Cc @zhaofengli @mdarocha 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

IvarWithoutBones committed:
baget: migrate lockfile generation to fetch-deps

IvarWithoutBones committed:
buildDotnetModule: restore for all platforms in fetch-deps

This makes buildDotnetModule restore nuget dependencies for the
platforms set in meta.platforms. This should help with generating
lockfiles for platforms other than the host machine.

Co-authored-by: mdarocha <git@mdarocha.pl>

IvarWithoutBones committed:
jackett: migrate lockfile generation to fetch-deps

IvarWithoutBones committed:
python-language-server: migrate lockfile generation to fetch-deps

IvarWithoutBones committed:
discordchatexporter-cli: migrate lockfile generation to fetch-deps

IvarWithoutBones committed:
opentabletdriver: migrate lockfile generation to fetch-deps

IvarWithoutBones committed:
osu-lazer: migrate lockfile generation to fetch-deps

IvarWithoutBones committed:
omnisharp-roslyn: migrate lockfile generation to fetch-deps

IvarWithoutBones commented on pkgs/build-support/dotnet/build-dotnet-module/default.nix:
```diff
@@ -138,25 +138,49 @@ in stdenvNoCC.mkDerivation (args // {
       cd "$(dirname "''${BASH_SOURCE[0]}")"
 
+      while (( $# > 0 )); do
+        case "$1" in
```
I wonder if it'd be useful to add an `--output` flag so we dont have any positional flags. Im a bit conflicted myself seeing how you'll basically always want to pass it. Curious what other people think.

IvarWithoutBones replied:
Not required anymore.

mdarocha commented:
Keep in mind that I have https://github.com/NixOS/nixpkgs/pull/179630 laying around, unfinished vecause I wanted to make the same changes you did here :).

You can probably copy the relevant `nuget-to-nix` here, which will allow you to replace the remaining scripts - both ArchiSteamFarm (where the arch emulation is not needed, dotnet will happily restore packages for another RID) and jellyfin should work properly then

mdarocha commented on pkgs/build-support/dotnet/build-dotnet-module/default.nix:
```diff
@@ -138,25 +138,49 @@ in stdenvNoCC.mkDerivation (args // {
 
+      while (( $# > 0 )); do
+        case "$1" in
+          --post-unpack)
```
Seeing what you use this flag for, I dont think its needed. I encountered a similar problem in `eventstore` - thats why I changed the update script to use `srcOnly`.

Then, if you put the post unpack changes in the `postPatch` hook, they will modify the source used by both the update script and the actual build. Then we have the added benefit of deduplication.

Unless I missed something, all usages of `--post-unpack` would be better with putting them in `postPatch`

IvarWithoutBones replied:
I had completely forgotten about that :sweat_smile: 

I reverted those changes, it indeed was not required at all. The only derivation that used it (`omnisharp-roslyn`) already worked out of the box.

IvarWithoutBones committed:
archisteamfarm: migrate update script to fetch-deps, regenerate deps

IvarWithoutBones committed:
jellyfin: migrate lockfile generation to fetch-deps

IvarWithoutBones commented:
> Keep in mind that I have #179630 laying around, unfinished because I wanted to make the same changes you did here :).

Interesting, I hadnt seen that PR before. I copied some of your code and switched over the remaining update scripts :+1: 

I noticed some packages do add `--runtime <RID>` flags themselves with `dotnetFlags` for example, which causes problems in the `fetch-deps` script because of duplicated RID flags. Setting the flag doesnt seem like a bad idea for some packages, for example ArchiSteamFarm attempts to build stuff for the windows target otherwise, which we obviously havent restored for.

Note that we cannot set the flag by default in `buildDotnetModule` to avoid other derivations needing to do so, some packages do not work well with it. Instead I've added some logic to filter out all flags containing RIDs, let me know what you think :)

IvarWithoutBones commented:
Note that ive regenerated the dependencies for archisteamfarm to merge the existing nuget lockfiles it used. Otherwise the update script couldnt be properly migrated. 

Now that the maintainers of that package are pinged anyway: I noticed it has disabled automated updates:
https://github.com/NixOS/nixpkgs/blob/858712163c522174181456edb008f3496b0a0561/pkgs/applications/misc/ArchiSteamFarm/default.nix#L14
 Is that because cross compilation used to be required for regenerating the dependency lockfile? And can we enable it again now?

IvarWithoutBones commented:
If no one objects I'll merge this in a week from now, in an attempt to keep momentum on dotnet build support going.

SuperSandro2000 commented on pkgs/applications/misc/ArchiSteamFarm/default.nix:
```diff
@@ -23,7 +23,10 @@ buildDotnetModule rec {
+  nugetDeps = ./deps.nix;
+
+  # Without this dotnet attempts to restore for Windows targets, which it cannot find the dependencies for
+  dotnetRestoreFlags = [ "--runtime linux-${if stdenvNoCC.hostPlatform.isAarch64 then "arm64" else "x64"}" ];
```
We shouldn't just fallback to x64. We should also check that and error if the platform is anything else.

zhaofengli replied:
We should probably have something similar to `toRustTarget` in our Rust infra to translate a platform name to the .Net Runtime ID.

SuperSandro2000 replied:
That also sounds like a good idea to reduce code repetition.

IvarWithoutBones replied:
> We should probably have something reusable similar to toRustTarget in our Rust infra to translate a platform name to the .Net Runtime ID.

I wonder what the right location for that would be. Maybe a passthru attribute from the SDK? We could also create a `dotnetPlatform` attribute, could include stuff like `mkNugetSource` as well.

IvarWithoutBones replied:
Added it to the SDK for now.

SuperSandro2000 commented on pkgs/build-support/dotnet/build-dotnet-module/default.nix:
```diff
@@ -135,43 +151,78 @@ in stdenvNoCC.mkDerivation (args // {
+        "aarch64-linux" = "linux-arm64";
+        "x86_64-darwin" = "osx-x64";
+        "aarch64-darwin" = "osx-arm64";
+      };
```
Lets add an throw unsupported system just to be safe.

IvarWithoutBones committed:
dotnet-sdk: add passthru attribute to generate a runtime ID

GGG-KILLER commented on pkgs/development/tools/omnisharp-roslyn/updater.sh:
```diff
@@ -1,19 +1,8 @@
 #! /usr/bin/env nix-shell
-#! nix-shell -i bash -p curl jq common-updater-scripts nuget-to-nix
+#! nix-shell -I nixpkgs=./. -i bash -p curl jq common-updater-scripts
```
Shouldn't this be `-I nixpkgs=../../../..`?
```suggestion
#! nix-shell -I nixpkgs=../../../.. -i bash -p curl jq common-updater-scripts
```

IvarWithoutBones replied:
IIRC the maintainer script will depend on running the update script from nixpkgs root. That's what r-ryantm uses, so compatibility is definitely something we want.

GGG-KILLER replied:
Oh ok, ignore all the other comments then

IvarWithoutBones replied:
Don't think `-I nixpkgs=./.` happens by default, using the current nixpkgs checkout is still valuable id say. Other comments still seem useful to me, just took some liberties with applying them :)

GGG-KILLER commented on pkgs/development/tools/omnisharp-roslyn/updater.sh:
```diff
@@ -29,38 +18,5 @@ fi
-done
 
-nuget-to-nix ./nuget_pkgs > "$deps_file"
+$(nix-build -A omnisharp-roslyn.fetch-deps --no-out-link) "$deps_file"
```
Not sure if this has been tested, but omnisharp has some .NET 7 dependencies (no idea why. although it can be compiled and ran on .NET 6, it cannot be restored on 7)

IvarWithoutBones replied:
Worked fine from my testing.

GGG-KILLER replied:
Oh ok, guess that was some weird compatibility issue I had with all the changes I was making then

GGG-KILLER commented on pkgs/development/dotnet-modules/python-language-server/updater.sh:
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env nix-shell
-#!nix-shell -i bash -p gnused jq common-updater-scripts nuget-to-nix dotnet-sdk_3 nix-prefetch-git
+#!nix-shell -i bash -p gnused jq common-updater-scripts nix-prefetch-git
```
This seems to be missing `-I nixpkgs=../../../..`?
```suggestion
#!nix-shell -I nixpkgs=../../../.. -i bash -p gnused jq common-updater-scripts nix-prefetch-git
```

GGG-KILLER commented on pkgs/servers/jackett/updater.sh:
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env nix-shell
-#!nix-shell -i bash -p curl jq common-updater-scripts nuget-to-nix dotnet-sdk_6
+#!nix-shell -i bash -p curl jq common-updater-scripts
```
This seems to be missing `-I nixpkgs=../../..`?
```suggestion
#!nix-shell -I nixpkgs=../../.. -i bash -p curl jq common-updater-scripts
```

GGG-KILLER commented on pkgs/servers/jellyfin/update.sh:
```diff
@@ -1,11 +1,8 @@
 #!/usr/bin/env nix-shell
-#!nix-shell -i bash -p curl jq common-updater-scripts dotnetCorePackages.sdk_6_0 nuget-to-nix gnused nix coreutils findutils
+#!nix-shell -i bash -p curl jq common-updater-scripts gnused nix coreutils
```
This seems to be missing `-I nixpkgs=../../..`?
```suggestion
#!nix-shell -I nixpkgs=../../.. -i bash -p curl jq common-updater-scripts gnused nix coreutils
```

GGG-KILLER commented on pkgs/servers/web-apps/baget/updater.sh:
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env nix-shell
-#!nix-shell -i bash -p gnused jq common-updater-scripts nuget-to-nix dotnet-sdk_3 nix-prefetch-github
+#!nix-shell -i bash -p gnused jq common-updater-scripts nix-prefetch-github
```
This seems to be missing `-I nixpkgs=../../../..`?
```suggestion
#!nix-shell -I nixpkgs=../../../.. -i bash -p gnused jq common-updater-scripts nix-prefetch-github
```

GGG-KILLER commented on pkgs/tools/X11/opentabletdriver/update.sh:
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env nix-shell
-#!nix-shell -i bash -p curl gnused jq common-updater-scripts nuget-to-nix dotnet-sdk_6 dotnet-sdk_5
+#!nix-shell -i bash -p curl gnused jq common-updater-scripts
```
This seems to be missing `-I nixpkgs=../../../..`?
```suggestion
#!nix-shell -I nixpkgs=../../../.. -i bash -p curl gnused jq common-updater-scripts
```

GGG-KILLER commented on pkgs/tools/backup/discordchatexporter-cli/updater.sh:
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env nix-shell
-#!nix-shell -i bash -p curl jq common-updater-scripts nuget-to-nix dotnet-sdk_6
+#!nix-shell -i bash -p curl jq common-updater-scripts
```
This seems to be missing `-I nixpkgs=../../../..`?
```suggestion
#!nix-shell -I nixpkgs=../../../.. -i bash -p curl jq common-updater-scripts
```

IvarWithoutBones commented:
@GGG-KILLER I applied your suggestions, though instead of `-I nixpkgs=../../..` I used `-I nixpkgs=./.` for all packages as explained in https://github.com/NixOS/nixpkgs/pull/187359#discussion_r965607996.

IvarWithoutBones commented:
Now that we throw when generating `fetch-deps` if it encounters a platform that we dont recognize, ive set the default platform to the one from the dotnet SDK. Removed occurances of `platforms.all` so everything still works fine.

SuperSandro2000 commented:
> Also fixed our usage of `mktemp` in `fetch-deps`, all our temporary files/directories now have a proper name and we dont hardcode `/tmp` anymore.

Do all packages also have a trap to clean up those temporary files afterwards?

IvarWithoutBones commented:
> > Also fixed our usage of `mktemp` in `fetch-deps`, all our temporary files/directories now have a proper name and we dont hardcode `/tmp` anymore.
> 
> Do all packages also have a trap to clean up those temporary files afterwards?

That's not required since the `fetch-deps` script only generates the lockfile checked into nixpkgs, it already takes care of deleting all temporary files for you.

Only exception is the case theres no explicit path set to push the lockfile to, but no update scripts do that.

IvarWithoutBones commented:
Applied https://github.com/NixOS/nixpkgs/pull/190403#discussion_r966861324 and https://github.com/NixOS/nixpkgs/pull/190403#discussion_r966863859. Is there anything blocking this? From my perspective this should be ready to get merged.

GGG-KILLER commented:
> 

I believe you need to wrap the `throw` in parenthesis (I did the same in #183385)

IvarWithoutBones commented:
> I believe you need to wrap the throw in parenthesis @IvarWithoutBones (I did the same in https://github.com/NixOS/nixpkgs/pull/183385)

Ah, I wasn't aware `throw` returns a string. Done :+1:

IvarWithoutBones commented:
Merging this later today.

IvarWithoutBones commented:
Generated lockfiles for a few packages on both `x86_64-linux` and `x86_64-darwin`, works fine.

Result of `nixpkgs-review pr 187359` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ArchiSteamFarm</li>
  </ul>
</details>
